conu/backend/nspawn/image.py

```diff
@@ -0,0 +1,332 @@
+                                                          SEP=self.special_separator,
+                                                          TAGS=self.tag)
+        return self._id
+
+    def is_present(self):
+        return self.get_id() in NspawnImage.list_all()
+
+    def pull(self):
+        """
+        Pull this image from URL. Raises an exception if the image is not found in
+        the registry.
+
+        :return: None
+        """
+        ident = self.get_id()
+        try:
+            if self._is_local():
+                logger.debug("Try to pull local file: {} -> {}".format(self.name, ident))
+                run_cmd(["machinectl", "--verify=no", "import-raw", self.name, ident])
+            else:
+                logger.debug("Try to pull URL: {} -> {}".format(self.name, ident))
+                print (" ".join(["machinectl", "--verify=no",  "pull-raw", self.name, ident]))
+                run_cmd(["machinectl", "--verify=no",  "pull-raw", self.name, ident])
+        except ValueError as error:
+            raise ConuException("There was an error while pulling the image %s: %s", self.name, error)
+
+    @staticmethod
+    def list_all():
+        data = run_cmd(["machinectl", "list-images"], return_output=True)
+        output = []
+        for line in data.split("\n"):
+            stripped = line.strip()
+            if not stripped or stripped.startswith("NAME") or "No images" in line or "images listed" in line:
+                continue
+            splitted = stripped.split(" ", 1)
+            if "raw" in splitted[1] and constants.CONU_ARTIFACT_TAG in splitted[0]:
+                output.append(splitted[0])
+        return output
+
+    def create_snapshot(self, name, tag):
+        source = self.get_metadata()["Path"]
+        logger.debug("Create Snapshot: %s -> %s" % (source, name))
+        if name and tag:
+            output_tag = "{}:{}".format(name,tag)
+        else:
+            output_tag = name or tag
+        return self.__class__(repository=source, tag=output_tag)
+
+    def tag_image(self, tag):
+        return self.create_snapshot(name=None, tag=tag)
+
+    def inspect(self, refresh=True):
+        return self.get_metadata(refresh=refresh)
+
+    def get_metadata(self, refresh=True):
+        if refresh or not self._metadata:
+            ident = self._id or self.get_full_name()
+            if not ident:
+                raise ConuException("This image does not have a valid identifier.")
+            self._metadata = output_decode(run_cmd(
+                ["machinectl", "--output=json", "show-image", ident], return_output=True))
+        return self._metadata
+
+    def rmi(self, force=False, via_name=False):
+        return run_cmd(["machinectl", "remove", self.get_id()])
+
+    def mount(self, mount_point=None):
+        return NspawnImageFS(self, mount_point=mount_point)
+
+    def _is_killed(self, name):
```

Comment:
nit: for is_killed I would expect quick response yes/no. Insides of this method are rather wait_for_killed_status. Also Probe could help with the check.

Reply:
good point, renamed, using probes could be changed in fututure, I'll add there TODO's